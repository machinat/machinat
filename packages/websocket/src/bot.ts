import type {
  MachinatNode,
  MachinatBot,
  MachinatUser,
  InitScopeFn,
  DispatchWrapper,
} from '@machinat/core';
import Engine from '@machinat/core/engine';
import Renderer from '@machinat/core/renderer';
import Queue from '@machinat/core/queue';
import ModuleUtilitiesI from '@machinat/core/base/ModuleUtilities';
import { makeClassProvider } from '@machinat/core/service';

import { WEBSOCKET } from './constant';
import { PlatformUtilitiesI } from './interface';
import { ServerP } from './server';
import {
  WebSocketConnection,
  WebSocketTopicChannel,
  WebSocketUserChannel,
} from './channel';
import createJobs from './utils/createJobs';
import WebSocketWorker from './worker';
import type {
  EventInput,
  WebSocketJob,
  WebSocketResult,
  WebSocketComponent,
  WebSocketDispatchFrame,
  ConnIdentifier,
  WebSocketDispatchChannel,
  WebSocketDispatchResponse,
} from './types';

type SendResult = {
  connections: WebSocketConnection[];
};

const toConnection = ({ serverId, id }: ConnIdentifier) =>
  new WebSocketConnection(serverId, id);

/**
 * @category Provider
 */
export class WebSocketBot
  implements
    MachinatBot<WebSocketDispatchChannel, WebSocketJob, WebSocketResult>
{
  private _server: ServerP<any, unknown>;
  engine: Engine<
    WebSocketDispatchChannel,
    EventInput,
    WebSocketComponent,
    WebSocketJob,
    WebSocketResult
  >;

  platform = WEBSOCKET;

  constructor(
    server: ServerP<any, unknown>,
    initScope?: InitScopeFn,
    dispatchWrapper?: DispatchWrapper<
      WebSocketJob,
      WebSocketDispatchFrame,
      WebSocketResult
    >
  ) {
    this._server = server;

    const queue = new Queue<WebSocketJob, WebSocketResult>();
    const worker = new WebSocketWorker(server);

    const renderer = new Renderer<EventInput, WebSocketComponent>(
      WEBSOCKET,
      () => {
        throw new TypeError(
          'general component not supported at websocket platform'
        );
      }
    );

    this.engine = new Engine(
      WEBSOCKET,
      renderer,
      queue,
      worker,
      initScope,
      dispatchWrapper
    );
  }

  async start(): Promise<void> {
    await this._server.start();
    this.engine.start();
  }

  async stop(): Promise<void> {
    await this._server.stop();
    this.engine.stop();
  }

  render(
    channel: WebSocketDispatchChannel,
    message: MachinatNode
  ): Promise<null | WebSocketDispatchResponse> {
    return this.engine.render<WebSocketDispatchChannel>(
      channel,
      message,
      createJobs
    );
  }

  async send(
    channel: WebSocketConnection,
    content: EventInput | EventInput[]
  ): Promise<SendResult> {
    const {
      results: [{ connections }],
    } = await this.engine.dispatchJobs(channel, [
      {
        target: channel,
        values: Array.isArray(content) ? content : [content],
      },
    ]);

    return { connections: connections.map(toConnection) };
  }

  async sendUser(
    user: MachinatUser,
    content: EventInput | EventInput[]
  ): Promise<SendResult> {
    const channel = new WebSocketUserChannel(user.uid);
    const {
      results: [{ connections }],
    } = await this.engine.dispatchJobs(channel, [
      {
        target: channel,
        values: Array.isArray(content) ? content : [content],
      },
    ]);

    return { connections: connections.map(toConnection) };
  }

  async sendTopic(
    topic: string,
    content: EventInput | EventInput[]
  ): Promise<SendResult> {
    const channel = new WebSocketTopicChannel(topic);
    const {
      results: [{ connections }],
    } = await this.engine.dispatchJobs(channel, [
      {
        target: channel,
        values: Array.isArray(content) ? content : [content],
      },
    ]);

    return { connections: connections.map(toConnection) };
  }

  disconnect(
    connection: WebSocketConnection,
    reason?: string
  ): Promise<boolean> {
    return this._server.disconnect(connection, reason);
  }

  subscribeTopic(
    connection: WebSocketConnection,
    topic: string
  ): Promise<boolean> {
    return this._server.subscribeTopic(connection, topic);
  }

  unsubscribeTopic(
    connection: WebSocketConnection,
    topic: string
  ): Promise<boolean> {
    return this._server.unsubscribeTopic(connection, topic);
  }
}

export const BotP = makeClassProvider({
  lifetime: 'singleton',
  deps: [
    ServerP,
    { require: ModuleUtilitiesI, optional: true },
    { require: PlatformUtilitiesI, optional: true },
  ],
  factory: (server, moduleUtils, platformUtils) =>
    new WebSocketBot(
      server,
      moduleUtils?.initScope,
      platformUtils?.dispatchWrapper
    ),
})(WebSocketBot);

export type BotP = WebSocketBot;
