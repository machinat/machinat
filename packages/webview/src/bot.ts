import type {
  MachinatNode,
  MachinatBot,
  MachinatUser,
  InitScopeFn,
  DispatchWrapper,
} from '@machinat/core';
import Engine from '@machinat/core/engine';
import Renderer from '@machinat/core/renderer';
import Queue from '@machinat/core/queue';
import { makeClassProvider } from '@machinat/core/service';
import ModuleUtilitiesI from '@machinat/core/base/ModuleUtilities';
import type { DispatchResponse } from '@machinat/core/engine';
import type { AnyServerAuthenticator } from '@machinat/auth';
import { WebSocketWorker } from '@machinat/websocket';
import { createJobs } from '@machinat/websocket/utils';
import type {
  EventInput,
  WebSocketJob,
  WebSocketResult,
  ConnIdentifier,
} from '@machinat/websocket';
import { WEBVIEW } from './constant';
import { SocketServerP, PlatformUtilitiesI } from './interface';
import {
  WebviewConnection,
  WebviewTopicChannel,
  WebviewUserChannel,
} from './channel';
import type {
  WebviewDispatchFrame,
  WebviewComponent,
  WebviewDispatchChannel,
} from './types';

type WebSocketDispatchResponse = DispatchResponse<
  WebSocketJob,
  WebSocketResult
>;

type SendResult = {
  connections: WebviewConnection[];
};

const toConnection = ({ serverId, id }: ConnIdentifier): WebviewConnection =>
  new WebviewConnection(serverId, id);

/**
 * @category Provider
 */
export class WebviewBot<Authenticator extends AnyServerAuthenticator>
  implements MachinatBot<WebviewDispatchChannel, WebSocketJob, WebSocketResult>
{
  private _server: SocketServerP<Authenticator>;

  engine: Engine<
    WebviewDispatchChannel,
    EventInput,
    WebviewComponent,
    WebSocketJob,
    WebSocketResult
  >;

  constructor(
    server: SocketServerP<Authenticator>,
    initScope?: InitScopeFn,
    dispatchWrapper?: DispatchWrapper<
      WebSocketJob,
      WebviewDispatchFrame,
      WebSocketResult
    >
  ) {
    this._server = server;

    const queue = new Queue<WebSocketJob, WebSocketResult>();
    const worker = new WebSocketWorker(server);

    const renderer = new Renderer<EventInput, WebviewComponent>(WEBVIEW, () => {
      throw new TypeError(
        'general component not supported at websocket platform'
      );
    });

    this.engine = new Engine(
      WEBVIEW,
      renderer,
      queue,
      worker,
      initScope,
      dispatchWrapper
    );
  }

  async start(): Promise<void> {
    await this._server.start();
    this.engine.start();
  }

  async stop(): Promise<void> {
    await this._server.stop();
    this.engine.stop();
  }

  render(
    channel: WebviewDispatchChannel,
    message: MachinatNode
  ): Promise<null | WebSocketDispatchResponse> {
    return this.engine.render<WebviewDispatchChannel>(
      channel,
      message,
      createJobs
    );
  }

  async send(
    channel: WebviewConnection,
    content: EventInput | EventInput[]
  ): Promise<SendResult> {
    const response = await this.engine.dispatchJobs(channel, [
      { target: channel, values: Array.isArray(content) ? content : [content] },
    ]);

    return {
      connections: response.results[0].connections.map(toConnection),
    };
  }

  async sendUser(
    user: MachinatUser,
    content: EventInput | EventInput[]
  ): Promise<SendResult> {
    const channel = new WebviewUserChannel(user.uid);
    const response = await this.engine.dispatchJobs(channel, [
      { target: channel, values: Array.isArray(content) ? content : [content] },
    ]);

    return {
      connections: response.results[0].connections.map(toConnection),
    };
  }

  async sendTopic(
    topic: string,
    content: EventInput | EventInput[]
  ): Promise<SendResult> {
    const channel = new WebviewTopicChannel(topic);
    const response = await this.engine.dispatchJobs(channel, [
      { target: channel, values: Array.isArray(content) ? content : [content] },
    ]);

    return {
      connections: response.results[0].connections.map(toConnection),
    };
  }

  disconnect(connection: WebviewConnection, reason?: string): Promise<boolean> {
    return this._server.disconnect(connection, reason);
  }

  subscribeTopic(
    connection: WebviewConnection,
    topic: string
  ): Promise<boolean> {
    return this._server.subscribeTopic(connection, topic);
  }

  unsubscribeTopic(
    connection: WebviewConnection,
    topic: string
  ): Promise<boolean> {
    return this._server.unsubscribeTopic(connection, topic);
  }
}

export const BotP = makeClassProvider({
  lifetime: 'singleton',
  deps: [
    SocketServerP,
    { require: ModuleUtilitiesI, optional: true },
    { require: PlatformUtilitiesI, optional: true },
  ] as const,
  factory: (server, moduleUitils, platformUtils) =>
    new WebviewBot(
      server,
      moduleUitils?.initScope,
      platformUtils?.dispatchWrapper
    ),
})(WebviewBot);

export type BotP<Authenticator extends AnyServerAuthenticator> =
  WebviewBot<Authenticator>;
