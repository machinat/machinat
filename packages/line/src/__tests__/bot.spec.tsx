import nock from 'nock';
import moxy, { Moxy } from '@moxyjs/moxy';
import Machinat from '@machinat/core';
import _Engine from '@machinat/core/engine';
import _Renderer from '@machinat/core/renderer';
import Queue from '@machinat/core/queue';
import { LineBot } from '../bot';
import LineWorker from '../worker';
import LineApiError from '../error';
import {
  Expression,
  Image,
  QuickReply,
  MessageAction,
  LinkRichMenu,
} from '../components';

const Renderer = _Renderer as Moxy<typeof _Renderer>;
const Engine = _Engine as Moxy<typeof _Engine>;

jest.mock('@machinat/core/engine', () =>
  jest
    .requireActual('@moxyjs/moxy')
    .default(jest.requireActual('@machinat/core/engine'))
);

jest.mock('@machinat/core/renderer', () =>
  jest
    .requireActual('@moxyjs/moxy')
    .default(jest.requireActual('@machinat/core/renderer'))
);

nock.disableNetConnect();

const message = (
  <Expression
    quickReplies={
      <QuickReply>
        <MessageAction text="Hi!" label="HI" />
      </QuickReply>
    }
  >
    Hello <b>LINE</b>
    <Image originalContentUrl="https://..." previewImageUrl="https://..." />
    <LinkRichMenu id="newbie" />
  </Expression>
);

const scope = moxy();
const initScope = moxy(() => scope);
const dispatchWrapper = moxy((x) => x);

let lineApi;
beforeEach(() => {
  Engine.mock.reset();
  Renderer.mock.reset();

  nock.cleanAll();
  lineApi = nock('https://api.line.me', {
    reqheaders: {
      'content-type': 'application/json',
      authorization: 'Bearer _ACCESS_TOKEN_',
    },
  });
});

describe('#constructor(options)', () => {
  it('throw if configs.channelId is empty', () => {
    expect(
      () =>
        new LineBot({
          providerId: '_PROVIDER_ID_',
          accessToken: '_ACCESS_TOKEN_',
        } as never)
    ).toThrowErrorMatchingInlineSnapshot(
      `"configs.channelId should not be empty"`
    );
  });

  it('throw if configs.providerId is empty', () => {
    expect(
      () =>
        new LineBot({
          channelId: '_CHANNEL_ID_',
          accessToken: '_ACCESS_TOKEN_',
        } as never)
    ).toThrowErrorMatchingInlineSnapshot(
      `"configs.providerId should not be empty"`
    );
  });

  it('throws if accessToken not given', () => {
    expect(
      () =>
        new LineBot({
          providerId: '_PROVIDER_ID_',
          channelId: '_CHANNEL_ID_',
        } as never)
    ).toThrowErrorMatchingInlineSnapshot(
      `"configs.accessToken should not be empty"`
    );
  });

  it('assemble engine', () => {
    const bot = new LineBot({
      accessToken: '_ACCESS_TOKEN_',
      providerId: '_PROVIDER_ID_',
      channelId: '_CHANNEL_ID_',
      maxConnections: 999,
      initScope,
      dispatchWrapper,
    });

    expect(Renderer.mock).toHaveBeenCalledTimes(1);
    expect(Renderer.mock).toHaveBeenCalledWith('line', expect.any(Function));

    expect(bot.engine).toBeInstanceOf(Engine);
    expect(Engine.mock).toHaveBeenCalledTimes(1);
    expect(Engine.mock).toHaveBeenCalledWith(
      'line',
      expect.any(Renderer),
      expect.any(Queue),
      expect.any(LineWorker),
      initScope,
      dispatchWrapper
    );

    const worker = Engine.mock.calls[0].args[3];
    expect(worker.accessToken).toBe('_ACCESS_TOKEN_');
    expect(worker.maxConnections).toBe(999);
  });
});

describe('#render(token, node, options)', () => {
  it('make api calls', async () => {
    const bot = new LineBot({
      accessToken: '_ACCESS_TOKEN_',
      providerId: '_PROVIDER_ID_',
      channelId: '_CHANNEL_ID_',
    });

    bot.start();

    const apiCall1 = lineApi
      .post('/v2/bot/message/push', {
        to: 'john_doe',
        messages: [
          { text: 'Hello LINE', type: 'text' },
          {
            type: 'image',
            originalContentUrl: 'https://...',
            previewImageUrl: 'https://...',
            quickReply: {
              items: [
                {
                  action: { label: 'HI', text: 'Hi!', type: 'message' },
                  type: 'action',
                },
              ],
            },
          },
        ],
      })
      .reply(200, '{}');
    const apiCall2 = lineApi
      .post('/v2/bot/user/john_doe/richmenu/newbie', '')
      .reply(200, '{}');

    const response = await bot.render('john_doe', message);

    expect(response).toMatchSnapshot();
    expect(apiCall1.isDone()).toBe(true);
    expect(apiCall2.isDone()).toBe(true);
  });

  it('works with replyToken', async () => {
    const bot = new LineBot({
      accessToken: '_ACCESS_TOKEN_',
      providerId: '_PROVIDER_ID_',
      channelId: '_CHANNEL_ID_',
    });
    bot.start();

    const apiCall1 = lineApi
      .post('/v2/bot/message/reply', {
        replyToken: '__REPLY_TOKEN__',
        messages: [
          { text: 'Hello LINE', type: 'text' },
          {
            type: 'image',
            originalContentUrl: 'https://...',
            previewImageUrl: 'https://...',
            quickReply: {
              items: [
                {
                  type: 'action',
                  action: { label: 'HI', text: 'Hi!', type: 'message' },
                },
              ],
            },
          },
        ],
      })
      .reply(200, '{}');

    const apiCall2 = lineApi
      .post('/v2/bot/user/john_doe/richmenu/newbie', '')
      .reply(200, '{}');

    const response = await bot.render('john_doe', message, {
      replyToken: '__REPLY_TOKEN__',
    });

    expect(response).toMatchSnapshot();
    expect(apiCall1.isDone()).toBe(true);
    expect(apiCall2.isDone()).toBe(true);
  });

  it('return null if message is empty', async () => {
    const bot = new LineBot({
      accessToken: '_ACCESS_TOKEN_',
      providerId: '_PROVIDER_ID_',
      channelId: '_CHANNEL_ID_',
    });

    for (const empty of [null, undefined, [], <></>, true, false]) {
      // eslint-disable-next-line no-await-in-loop
      await expect(bot.render('john_doe', empty)).resolves.toBe(null);
    }
  });
});

describe('#renderMulticast(targets, node)', () => {
  it('return null if message is empty', async () => {
    const bot = new LineBot({
      accessToken: '_ACCESS_TOKEN_',
      providerId: '_PROVIDER_ID_',
      channelId: '_CHANNEL_ID_',
    });

    for (const empty of [null, undefined, [], <></>, true, false]) {
      // eslint-disable-next-line no-await-in-loop
      await expect(
        bot.renderMulticast(['no', 'one', 'knows'], empty)
      ).resolves.toBe(null);
    }
  });

  it('make api call to message/mulitcast', async () => {
    const bot = new LineBot({
      accessToken: '_ACCESS_TOKEN_',
      providerId: '_PROVIDER_ID_',
      channelId: '_CHANNEL_ID_',
    });
    bot.start();

    const apiCall1 = lineApi
      .post('/v2/bot/message/multicast', {
        to: ['john', 'wick', 'dog'],
        messages: [
          { text: 'Hello LINE', type: 'text' },
          {
            type: 'image',
            originalContentUrl: 'https://...',
            previewImageUrl: 'https://...',
            quickReply: {
              items: [
                {
                  type: 'action',
                  action: { label: 'HI', text: 'Hi!', type: 'message' },
                },
              ],
            },
          },
        ],
      })
      .reply(200, '{}');

    const apiCall2 = lineApi
      .post('/v2/bot/richmenu/bulk/link', {
        userIds: ['john', 'wick', 'dog'],
        richMenuId: 'newbie',
      })
      .reply(200, '{}');

    const response = await bot.renderMulticast(
      ['john', 'wick', 'dog'],
      message
    );

    expect(response).toMatchSnapshot();
    expect(apiCall1.isDone()).toBe(true);
    expect(apiCall2.isDone()).toBe(true);
  });
});

describe('#makeApiCall()', () => {
  it('call line REST api', async () => {
    const bot = new LineBot({
      accessToken: '_ACCESS_TOKEN_',
      providerId: '_PROVIDER_ID_',
      channelId: '_CHANNEL_ID_',
    });
    bot.start();

    const apiCall = lineApi
      .post('/v2/bot/foo', { bar: 'baz' })
      .reply(200, '{"foo":"bar.baz"}');

    await expect(
      bot.makeApiCall('POST', 'v2/bot/foo', { bar: 'baz' })
    ).resolves.toEqual({
      foo: 'bar.baz',
    });

    expect(apiCall.isDone()).toBe(true);
  });

  it('throw LineApiError if api call fail', async () => {
    const bot = new LineBot({
      accessToken: '_ACCESS_TOKEN_',
      providerId: '_PROVIDER_ID_',
      channelId: '_CHANNEL_ID_',
    });
    bot.start();

    const apiCall = lineApi
      .post('/v2/bot/foo', { bar: 'baz' })
      .reply(400, { message: 'bad' });

    try {
      await bot.makeApiCall('POST', 'v2/bot/foo', { bar: 'baz' });
      expect('should not be here').toBeFalsy();
    } catch (err) {
      expect(err).toBeInstanceOf(LineApiError);
      expect(err.code).toBe(400);
      expect(err.info).toEqual({ message: 'bad' });
    }

    expect(apiCall.isDone()).toBe(true);
  });
});
