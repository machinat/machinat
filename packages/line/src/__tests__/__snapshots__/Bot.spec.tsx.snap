// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render(token, node, options) make api calls 1`] = `
Object {
  "jobs": Array [
    Object {
      "body": Object {
        "messages": Array [
          Object {
            "text": "Hello LINE",
            "type": "text",
          },
          Object {
            "originalContentUrl": "https://...",
            "previewImageUrl": "https://...",
            "quickReply": Object {
              "items": Array [
                Object {
                  "action": Object {
                    "label": "HI",
                    "text": "Hi!",
                    "type": "message",
                  },
                  "imageUrl": undefined,
                  "type": "action",
                },
              ],
            },
            "type": "image",
          },
        ],
        "to": "john_doe",
      },
      "executionKey": "line._CHANNEL_ID_.john_doe",
      "method": "POST",
      "path": "v2/bot/message/push",
    },
    Object {
      "body": null,
      "executionKey": "line._CHANNEL_ID_.john_doe",
      "method": "POST",
      "path": "v2/bot/user/john_doe/richmenu/newbie",
    },
  ],
  "results": Array [
    Object {
      "body": Object {},
      "code": 200,
      "headers": Object {},
    },
    Object {
      "body": Object {},
      "code": 200,
      "headers": Object {},
    },
  ],
  "tasks": Array [
    Object {
      "payload": Array [
        Object {
          "body": Object {
            "messages": Array [
              Object {
                "text": "Hello LINE",
                "type": "text",
              },
              Object {
                "originalContentUrl": "https://...",
                "previewImageUrl": "https://...",
                "quickReply": Object {
                  "items": Array [
                    Object {
                      "action": Object {
                        "label": "HI",
                        "text": "Hi!",
                        "type": "message",
                      },
                      "imageUrl": undefined,
                      "type": "action",
                    },
                  ],
                },
                "type": "image",
              },
            ],
            "to": "john_doe",
          },
          "executionKey": "line._CHANNEL_ID_.john_doe",
          "method": "POST",
          "path": "v2/bot/message/push",
        },
        Object {
          "body": null,
          "executionKey": "line._CHANNEL_ID_.john_doe",
          "method": "POST",
          "path": "v2/bot/user/john_doe/richmenu/newbie",
        },
      ],
      "type": "dispatch",
    },
  ],
}
`;

exports[`#render(token, node, options) works with replyToken 1`] = `
Object {
  "jobs": Array [
    Object {
      "body": Object {
        "messages": Array [
          Object {
            "text": "Hello LINE",
            "type": "text",
          },
          Object {
            "originalContentUrl": "https://...",
            "previewImageUrl": "https://...",
            "quickReply": Object {
              "items": Array [
                Object {
                  "action": Object {
                    "label": "HI",
                    "text": "Hi!",
                    "type": "message",
                  },
                  "imageUrl": undefined,
                  "type": "action",
                },
              ],
            },
            "type": "image",
          },
        ],
        "replyToken": "__REPLY_TOKEN__",
      },
      "executionKey": "line._CHANNEL_ID_.john_doe",
      "method": "POST",
      "path": "v2/bot/message/reply",
    },
    Object {
      "body": null,
      "executionKey": "line._CHANNEL_ID_.john_doe",
      "method": "POST",
      "path": "v2/bot/user/john_doe/richmenu/newbie",
    },
  ],
  "results": Array [
    Object {
      "body": Object {},
      "code": 200,
      "headers": Object {},
    },
    Object {
      "body": Object {},
      "code": 200,
      "headers": Object {},
    },
  ],
  "tasks": Array [
    Object {
      "payload": Array [
        Object {
          "body": Object {
            "messages": Array [
              Object {
                "text": "Hello LINE",
                "type": "text",
              },
              Object {
                "originalContentUrl": "https://...",
                "previewImageUrl": "https://...",
                "quickReply": Object {
                  "items": Array [
                    Object {
                      "action": Object {
                        "label": "HI",
                        "text": "Hi!",
                        "type": "message",
                      },
                      "imageUrl": undefined,
                      "type": "action",
                    },
                  ],
                },
                "type": "image",
              },
            ],
            "replyToken": "__REPLY_TOKEN__",
          },
          "executionKey": "line._CHANNEL_ID_.john_doe",
          "method": "POST",
          "path": "v2/bot/message/reply",
        },
        Object {
          "body": null,
          "executionKey": "line._CHANNEL_ID_.john_doe",
          "method": "POST",
          "path": "v2/bot/user/john_doe/richmenu/newbie",
        },
      ],
      "type": "dispatch",
    },
  ],
}
`;

exports[`#renderMulticast(targets, node) make api call to message/mulitcast 1`] = `
Object {
  "jobs": Array [
    Object {
      "body": Object {
        "messages": Array [
          Object {
            "text": "Hello LINE",
            "type": "text",
          },
          Object {
            "originalContentUrl": "https://...",
            "previewImageUrl": "https://...",
            "quickReply": Object {
              "items": Array [
                Object {
                  "action": Object {
                    "label": "HI",
                    "text": "Hi!",
                    "type": "message",
                  },
                  "imageUrl": undefined,
                  "type": "action",
                },
              ],
            },
            "type": "image",
          },
        ],
        "to": Array [
          "john",
          "wick",
          "dog",
        ],
      },
      "executionKey": "$$_multicast_$$",
      "method": "POST",
      "path": "v2/bot/message/multicast",
    },
    Object {
      "body": Object {
        "richMenuId": "newbie",
        "userIds": Array [
          "john",
          "wick",
          "dog",
        ],
      },
      "executionKey": "$$_multicast_$$",
      "method": "POST",
      "path": "v2/bot/richmenu/bulk/link",
    },
  ],
  "results": Array [
    Object {
      "body": Object {},
      "code": 200,
      "headers": Object {},
    },
    Object {
      "body": Object {},
      "code": 200,
      "headers": Object {},
    },
  ],
  "tasks": Array [
    Object {
      "payload": Array [
        Object {
          "body": Object {
            "messages": Array [
              Object {
                "text": "Hello LINE",
                "type": "text",
              },
              Object {
                "originalContentUrl": "https://...",
                "previewImageUrl": "https://...",
                "quickReply": Object {
                  "items": Array [
                    Object {
                      "action": Object {
                        "label": "HI",
                        "text": "Hi!",
                        "type": "message",
                      },
                      "imageUrl": undefined,
                      "type": "action",
                    },
                  ],
                },
                "type": "image",
              },
            ],
            "to": Array [
              "john",
              "wick",
              "dog",
            ],
          },
          "executionKey": "$$_multicast_$$",
          "method": "POST",
          "path": "v2/bot/message/multicast",
        },
        Object {
          "body": Object {
            "richMenuId": "newbie",
            "userIds": Array [
              "john",
              "wick",
              "dog",
            ],
          },
          "executionKey": "$$_multicast_$$",
          "method": "POST",
          "path": "v2/bot/richmenu/bulk/link",
        },
      ],
      "type": "dispatch",
    },
  ],
}
`;
